File: clients/cpp/templates/RemoteAPIObjects.cpp
#py from calltip import FuncDef
#py from cpp_utils import *
#py all_func_defs = FuncDef.from_calltips_json(pycpp.params['calltips_json'], pycpp.params['include_objects'], pycpp.params['exclude_objects'], pycpp.params['exclude_methods'])
namespace RemoteAPIObject
{
#py for obj, func_defs in all_func_defs.items():
    `obj`::`obj`(RemoteAPIClient *client)
        : _client(client)
    {
        _client->require("`obj`");
    }

#py if obj == 'sim':
#include "sim-deprecated.cpp"
#include "sim-special.cpp"

#py endif
#py for func, func_def in func_defs.items():
#py if func_def.in_args.is_variadic() or func_def.out_args.is_variadic():
#py continue
#py endif
    `cpp_rets(func_def.out_args)` `obj`::`func`(`cpp_args(func_def.in_args)`)
    {
        bool _brk = false;
        json _args(json_array_arg);
#py for arg in func_def.in_args:
#py wrap = (lambda x: f'bin({x})') if arg.type == 'buffer' else (lambda x: x)
#py if arg.is_optional():
        if(`arg.name`)
        {
            if(_brk) throw std::runtime_error("no gaps allowed");
            else _args.push_back(`wrap(f'*{arg.name}')`);
        }
        else _brk = true;
#py else:
        _args.push_back(`wrap(arg.name)`);
#py endif
#py endfor
        auto _ret = this->_client->call("`func_def.func_name`", _args);
#py if len(func_def.out_args) == 1:
        return _ret[0].as<`cpp_type(func_def.out_args[0])`>();
#py elif len(func_def.out_args) > 1:
        return std::make_tuple(`', '.join(f'_ret[{i}].as<{cpp_type(arg)}>()' for i, arg in enumerate(func_def.out_args))`);
#py endif
    }

#py endfor
#py endfor
};

RemoteAPIObjects::RemoteAPIObjects(RemoteAPIClient *client)
{
    this->client = client;
}

#py for obj, func_defs in all_func_defs.items():
RemoteAPIObject::`obj` RemoteAPIObjects::`obj`()
{
    return RemoteAPIObject::`obj`(this->client);
}

#py endfor

File: clients/java/templates/RemoteAPIObjects.java
#py from calltip import FuncDef
#py from java_utils import *
#py import json
#py all_func_defs = FuncDef.from_calltips_json(pycpp.params['calltips_json'], pycpp.params['include_objects'], pycpp.params['exclude_objects'], pycpp.params['exclude_methods'])
#py f = open(pycpp.params['constants_json'], 'rt')
#py all_constants = json.load(f)
#py f.close()
#py reservedNames = {'abstract', 'assert', 'boolean', 'break', 'byte', 'case', 'catch', 'char', 'class', 'continue', 'const', 'default', 'do', 'double', 'else', 'enum', 'exports', 'extends', 'final', 'finally', 'float', 'for', 'goto', 'if', 'implements', 'import', 'instanceof', 'int', 'interface', 'long', 'module', 'native', 'new', 'package', 'private', 'protected', 'public', 'requires', 'return', 'short', 'static', 'strictfp', 'super', 'switch', 'synchronized', 'this', 'throw', 'throws', 'transient', 'try', 'var', 'void', 'volatile', 'while'}
#py fixReserved = lambda n: n + ('_' if n in reservedNames else '')
package com.coppeliarobotics.remoteapi.zmq;

import java.util.*;

import co.nstant.in.cbor.*;

public class RemoteAPIObjects
{
    private final RemoteAPIClient client;

    public RemoteAPIObjects(RemoteAPIClient client)
    {
        this.client = client;
    }

#py for obj, func_defs in all_func_defs.items():
    public class _`obj` extends `f"com.coppeliarobotics.remoteapi.zmq.objects.special._{obj}" if obj in ('sim', ) else "RemoteAPIObject"`
    {
        public _`obj`(RemoteAPIClient client)
        {
            super(client);
        }

#py for func, func_def in func_defs.items():
#py if func_def.in_args.is_variadic() or func_def.out_args.is_variadic():
#py continue
#py endif
        public `java_rets(func_def.out_args)` `func`(Object... args) throws CborException
        {
#py if len(func_def.out_args) == 0:
            this.client.call("`obj`.`func`", args);
#py else:
            Object[] rets = this.client.call("`obj`.`func`", args);
#py if len(func_def.out_args) > 1:
            return rets;
#py else:
            return ((`java_type(func_def.out_args[0])`)rets[0]);
#py endif
#py endif
        }

#py endfor
#py for k, v in sorted(all_constants.get(obj, {}).items()):
#py if isinstance(v, dict):
        class _`k` {
#py for k1, v1 in sorted(v.items()):
            public static final int `fixReserved(k1)` = `v1`;
#py endfor
        }

        public static final _`k` `fixReserved(k)`;
#py else:
        public static final int `fixReserved(k)` = `v`;
#py endif
#py endfor
    }

    public _`obj` `obj`() {
        return new _`obj`(this.client);
    }

#py endfor
}
